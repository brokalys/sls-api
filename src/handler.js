import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (process.env.STAGE !== 'prod' && process.env.NODE_ENV !== 'test') ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (process.env.NODE_ENV !== 'test') {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
