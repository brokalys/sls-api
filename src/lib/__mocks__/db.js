export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
