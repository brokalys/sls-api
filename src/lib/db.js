import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
