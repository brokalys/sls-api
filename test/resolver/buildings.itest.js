import { createTestClient } from 'apollo-server-testing';
import { server } from 'handler';
import { authenticateAs } from 'helpers';
import db from 'db-config';

const { query } = createTestClient(server);

describe('Resolver: buildings', () => {
  beforeAll(async () => {
    await db.migrate.latest();
    await db.seed.run();
  });

  afterEach(jest.clearAllMocks);

  test('successfully retrieves single building', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query GetBuilding($id: Int!) {
          building(id: $id) {
            id
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves all the available fields', async () => {
    authenticateAs('slsCrawler', server);

    const response = await query({
      query: `
        query ($id: Int!) {
          building(id: $id) {
            id
            bounds
            properties {
              results {
                id
                url
                category
                type
                rent_type
                content
                images
                price
                price_per_sqm
                calc_price_per_sqm
                rooms
                area
                floor
                lat
                lng
                published_at
                foreign_id
              }
              summary {
                count
                price {
                  count
                  min
                  max
                  mean
                  median
                  mode
                  standardDev
                }
              }
            }
            vzd {
              apartments {
                results {
                  id
                  cadastre_number
                  property_address
                  sale_date
                  price
                  land_cadastral_designations
                  land_area_m2
                  building_depreciation_percentage
                  building_cadastral_designations
                  space_group_lowest_floor
                  space_group_highest_floor
                  apartment_total_area_m2
                  room_count
                }
              }
            }
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves building an property information', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query GetBuildingsAndProperties($id: Int!) {
          building(id: $id) {
            id
            properties {
              results {
                price
              }
            }
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves building VZD sales information', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query($id: Int!) {
          building(id: $id) {
            id
            vzd {
              apartments {
                results {
                  id
                  cadastre_number
                  price
                }
              }
            }
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('throws a validation exception if no id filter provided', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        {
          building {
            id
          }
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('throws an authentication error if trying to retrieve results without authorizing', async () => {
    const response = await query({
      query: `
        query GetBuildingsAndProperties($id: Int!) {
          building(id: $id) {
            id
            properties {
              results {
                price
              }
            }
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response.errors).toEqual([
      expect.objectContaining({
        extensions: {
          code: 'UNAUTHENTICATED',
        },
      }),
    ]);
  });

  test('throws an authentication error if trying to retrieve results without authorizing', async () => {
    const response = await query({
      query: `
        query($id: Int!) {
          building(id: $id) {
            id
            vzd {
              apartments {
                results {
                  price
                }
              }
            }
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response.errors).toEqual([
      expect.objectContaining({
        extensions: {
          code: 'UNAUTHENTICATED',
        },
      }),
    ]);
  });
});
