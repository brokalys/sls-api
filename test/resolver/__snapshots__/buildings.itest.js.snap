// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolver: buildings successfully retrieves all the available fields 1`] = `
Object {
  "data": Object {
    "building": Object {
      "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619",
      "id": 1,
      "properties": Object {
        "results": Array [
          Object {
            "area": 90,
            "calc_price_per_sqm": 1.11,
            "category": "apartment",
            "content": "",
            "floor": 2,
            "foreign_id": null,
            "id": 1,
            "images": Array [],
            "lat": 56.1,
            "lng": 24.9,
            "price": 100,
            "price_per_sqm": 1.11,
            "published_at": "2021-04-12T19:47:30.000Z",
            "rent_type": "monthly",
            "rooms": null,
            "type": "rent",
            "url": "https://www.ss.com/msg/lv/real-estate/flats/aizkraukle-and-reg/aiviekstes-pag/aghfx.html",
          },
          Object {
            "area": null,
            "calc_price_per_sqm": null,
            "category": "apartment",
            "content": "",
            "floor": null,
            "foreign_id": null,
            "id": 2,
            "images": Array [],
            "lat": 56.1,
            "lng": 24.9,
            "price": 200,
            "price_per_sqm": null,
            "published_at": "2021-04-12T19:47:30.000Z",
            "rent_type": "monthly",
            "rooms": null,
            "type": "rent",
            "url": "https://www.ss.lv/msg/lv/real-estate/flats/riga/bolderaya/cceeb.html",
          },
        ],
        "summary": Object {
          "count": 2,
          "price": Object {
            "count": 2,
            "max": 200,
            "mean": 150,
            "median": 150,
            "min": 100,
            "mode": 100,
            "standardDev": 50,
          },
        },
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: buildings successfully retrieves building an property information 1`] = `
Object {
  "data": Object {
    "building": Object {
      "id": 1,
      "properties": Object {
        "results": Array [
          Object {
            "price": 100,
          },
          Object {
            "price": 200,
          },
        ],
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: buildings successfully retrieves single building 1`] = `
Object {
  "data": Object {
    "building": Object {
      "id": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: buildings throws a validation exception if no id filter provided 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Field \\"building\\" argument \\"id\\" of type \\"Int!\\" is required, but it was not provided.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
